blockchain/state/statedb.go

```diff
@@ -541,34 +572,62 @@ func (self *StateDB) deleteStateObject(stateObject *stateObject) {
 	self.setError(self.trie.TryDelete(addr[:]))
 }
 
-// Retrieve a state object given by the address. Returns nil if not found.
+// getStateObject retrieves a state object given by the address, returning nil if
+// the object is not found or was deleted in this execution context. If you need
+// to differentiate between non-existent/just-deleted, use getDeletedStateObject.
 func (self *StateDB) getStateObject(addr common.Address) *stateObject {
+	if obj := self.getDeletedStateObject(addr); obj != nil && !obj.deleted {
+		return obj
+	}
+	return nil
+}
+
+// getDeletedStateObject is similar to getStateObject, but instead of returning
+// nil for a deleted state object, it returns the actual object with the deleted
+// flag set. This is needed by the state journal to revert to the correct s-
+// destructed object instead of wiping all knowledge about the state object.
+func (self *StateDB) getDeletedStateObject(addr common.Address) *stateObject {
 	// First, check stateObjects if there is "live" object.
 	if obj := self.stateObjects[addr]; obj != nil {
-		if obj.deleted {
-			return nil
-		}
 		return obj
 	}
-	// Track the amount of time wasted on loading the object from the database
-	if EnabledExpensive {
-		defer func(start time.Time) { self.AccountReads += time.Since(start) }(time.Now())
-	}
-	// Second, the object for given address is not cached.
-	// Load the object from the database.
-	enc, err := self.trie.TryGet(addr[:])
-	if len(enc) == 0 {
-		self.setError(err)
-		return nil
+	// If no live objects are available, attempt to use snapshots
+	var (
+		acc account.Account
+		err error
+	)
+	if self.snap != nil {
+		if EnabledExpensive {
+			defer func(start time.Time) { self.SnapshotAccountReads += time.Since(start) }(time.Now())
+		}
+		if acc, err = self.snap.Account(crypto.Keccak256Hash(addr.Bytes())); err == nil {
+			if acc == nil {
+				return nil
```

Comment:
It seems that snap.Account() returns nil if the account is not found.
Then don't we need to check the state trie?

Reply:
Should we? A snapshot contains all leaf nodes only based on the specific root hash. 
But, if an account is not found with an error, then it will try to get from state trie. Can this be the answer your question?

Reply:
okay. got it.